### ISSUE TITLE
Discovery -- insufficient metadata to sort creators by last name

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#1774]
In #1755, I pointed out that current creator metadata is insufficient to sort that metadata by last name. The reason for this is that a creator **who is not a HydroShare user** has just a "name", and not "first" and "last" names. HydroShare users have "first" and "last" names in their user profiles, but non-users just have a "name" field in the Party model, where Creator is a subclass of Party. 

The insufficiency of the Party model also came to light recently when someone listed a resource with no Creators and just an Organization. This crashed discovery. The fact that we can write non-sensical things into metadata is not acceptable. Meanwhile, there is no consistent mechanism for describing a creator that is an organization rather than a person. 

To address this problem, we can either "Band-Aid it" or really fix it. 

1. The Band-Aid is to create first and last name fields on the Party model that are inherited by Creator, Contributor, etc. 

2. The real fix is to rethink what it means to be a creator and come up with a consistent pattern for coding creators, which

    a. Creates **one record** for each non-HydroShare user that is reused for all of that user's 
        contributions.

    b. Allows that kind of user profile to be bound to a HydroShare user at a later time, when that person 
       gets an account. 

    c. Allows a current user to be removed without removing the user profile, so that people who are not 
       users can still be creators or contributors.  

    d. Disambiguates between "affiliation at time of authorship" and "current affiliation". 

    e. Defines a strict pattern for considering an organization to be a creator or contributor.

**Comment by dtarb** [comment]
Sort by last name is a new use case.  

This has been addressed (but perhaps in band aid fashion) in the rules for generation of citations, where there is the rule that a name can be expressed as
Sen Gupta, Avirup
or
Wilkins Jr., Pete Paul Mary
and then the entire part prior to the comma is interpreted as lastname.  For example the citation may be:

> Tarboton, D., A. Sen Gupta, P. P. M. Wilkins Jr. (2017). The Logan River Watershed, HydroShare, http://www.hydroshare.org/resource/GUID 

There must be code that is part of the citation functionality that identifies last name.  Could this be generalized to also use for last name sorting.  

With respect to the creator (and contributor) model, I think that when a resource is created, or an author or contributor is added to a resource, that the "party" record should be part of the resource with changes thereafter only by resource editing.  The act of filling information in from an existing HydroShare user is a convenience but should not formally link the resource to the user. We can retain a landing page URL link internally, that would go away if the user is deleted (We do not yet have account deletion functionality but need it). 

I do not think we should try to create a table with records for non-HydroShare users that are separate from resources. Managing such a table seems problematic to me, as no one has ownership of each record.  Thus I think that (a) "one record" for each non-HydroShare user is not a good idea. 

I do not think that (b) is really a need.

I think that (c) could be accommodated readily with our current model.  We have similar patterns implemented already with for example collections when a resource belonging to a collection is deleted, or relationships when a resource in a relationship is deleted.

I think that (d) is accommodated by the copy at time of creation and any changes thereafter be an edit action on the resource.

We may already have the strict patter of organization or person of (e).  The current interface for adding a creator or contributor allows setting of whether it is an organization or individual. I think that during the implementation of citation functionality there was code written to record for each creator/contributor whether they are an organization or Person.

**Comment by alvacouch** [comment]
The current REST API does not honor the UI restrictions on Creator. See issue #1788.

I disagree with the philosophy that the user can figure out the ambiguities above. They can be really difficult to figure out. 

There is a python library "[Probable People](https://parserator.datamade.us/probablepeople)" that does a much better -- though not perfect -- job of parsing peoples' names into first/middle/last. For a demo, click the link. This does better than our current bandaids. But I think we should hold ourselves to higher standards than that. Getting the citation approximately correct is not really acceptable. 

One easy way to address some of this is to allow the user to input the ORCId of the person, which disambiguates and parses the name appropriately through the ORCid API. If this is part of both the HydroShare user metadata and the Creator/Contributor metadata, then correct citation is assured.  See #1787

**Comment by hyi** [comment]
@alvacouch @pkdash @dtarb This may be a naive question, but why don't we ask users to input first name and last name (and middle name) when adding authors/contributors rather than just asking them to input whole name? Of course, for authors with identical names who are not hydroshare users, there will still be ambiguity which can only be resolved with an ID. ORCid is great, but not everyone has one yet. Also, do we really need to disambiguate each author who is not a hydroshare user? If we ask users to add first, middle, and last names rather than whole name when adding authors/contributors, we will form correct citation anyway.

**Comment by alvacouch** [comment]
@hyi I agree. But I have too much to do to make this happen right now.....

**Comment by Castronova** [comment]
related to #1021
